[DO NOT MERGE] Try for OAM-80992 random building issue

renchenglei opened the pull request:
We encounter vk random building issue on Android Q, please refer
to following error log.

https://buildbot.sh.intel.com/abspb/builders/build-q/builds/13888/steps/build_all/logs/stdio

renchenglei commented:
Please don't merge this PR, I put here for building test. We have one random vk building issue, which is only reproduced with buildbot. I have to create PR and build with buildbot.

tpalli commented:
It looks strange, libmesa_vulkan_util should come within libmesa_vulkan_common

renchenglei commented:
@tpalli, yes, the reproduced rate is very low. Not sure, if it is caused by building rule changes on Q. Maybe some sub-module is built parallel.

renchenglei commented:
@tpalli, I have updated the patch. With the new changes, we have tried 15 buildings. And issue seems gone. Do we still need those two flags on Android?

tpalli commented:
> @tpalli, I have updated the patch. With the new changes, we have tried 15 buildings. And issue seems gone. Do we still need those two flags on Android?

Yes we do need these flags, they are also used with i965 driver. Problem is that otherwise some other hash algorithm (default from toolchain) might get chosen and we explicitly want 'sha1'. I don't understand how having these flags could cause the linking issue. Maybe you were just lucky in this case?

renchenglei commented:
@tpalli, I tried to reproduce this issue on buildbot without any changes, and the reproduced rate is about 40%. Now with this changes, after 15 building, and no issue now.
If 'sha1' is must have. How about 'wl'?
The failure is about ld issues:
ld.lld: error: undefined symbol: vk_Result_to_str

tpalli commented:
> @tpalli, I tried to reproduce this issue on buildbot without any changes, and the reproduced rate is about 40%. Now with this changes, after 15 building, and no issue now.
> If 'sha1' is must have. How about 'wl'?

-Wl is the prefix used to give linker parameters when linker is run indirectly (via compiler), so it's not a parameter itself, just the indication that we want to give parameter '--build-id'

> The failure is about ld issues:
> ld.lld: error: undefined symbol: vk_Result_to_str

I do find it very strange if the linker parameter affects this. You could try if Vulkan apps work without these flags. When enabling Vulkan I had to put this flag, otherwise the build-id check would fail (see anv_physical_device_init_uuids in anv_device.c).

renchenglei commented:
@tpalli, that make sense! If possible, let's keep the origin patch. I updated the patch, I added "LOCAL_LDFLAGS += -Wl,--no-threads", which means we don't use multi-threads to do link. I triggered about 20 buildings, and not reproduced the random failure issue. Please take a review the latest patch. Thanks a lot!

tpalli commented:
All right, adding '--no-threads' is fine for me, it will mean slower linking but that will affect only the vulkan driver. It could possibly indicate a bug in Q toolchain as this has been working well so far.

renchenglei commented:
Thanks @tpalli. Will we keep this as a internal changes, if yes, I will update the PR commit info, and get it merged asap. As the random failure break Cactus team's auto building.

renchenglei commented:
@tpalli & @YuanjunHuang FYI.
With our previous changes(do not use multi-threads to link), we still could reproduce this issue. The reproduced rate is about 5%. It is so strange, this issue can't be reproduced on my local host, it is only reproduced on build bot, which could build with more than 30 processes.
There is some building enhancement on Android Q, it is much better not to generate *.h or *.c during building. This issue is caused by build&link parallel. Some files are not generated&compile before it is built&called.
I have check the source code, it seems that only one file call function vk_Result_to_str
```
grep -r vk_Result_to_str *
src/intel/vulkan/anv_util.c:   const char *error_str = vk_Result_to_str(error);
src/amd/vulkan/radv_util.c:     const char *error_str = vk_Result_to_str(error);
```
So I reorg the generation order, we hope it has been generated before we call it. I will trigger many many buildings with build bot to check if issue is still here.

renchenglei commented:
I tried to move the generated file code to libmesa_vulkan_common, but issue is still here, :(
https://buildbot.sh.intel.com/abspb/builders/build-q/builds/16187/steps/build_all/logs/stdio
It is so tough, could we copy the function to anv_util.c directly, which could help pass the building issue. And if yes, I suggested we only get it merged on 1A which won't block daily/weekly and integration build, we still could use github and opensource-gfx build bot to keep debugging this issue.

renchenglei commented:
@tpalli & @YuanjunHuang, what do you think of this changes, could we use this as a hot fix for cactus build?  
After many potential fixs, the vk random link failure is still here. Maybe we need much more time to investigate this issue, which is only reproduced on build bot with a very low reproduced rate.

tpalli commented:
Are you able to access the build machine? It would be interesting to see what symbols 'libmesa_vulkan_util.a' and 'libmesa_vulkan_common.a' have in them, you can list the symbols as example like: "readelf -s --wide libmesa_vulkan_common.a'"

tpalli commented:
Overall it looks to me that this is likely a bug in Android Q toolchain as the dependencies are set correctly in mk files. Using a workaround is ok for me but it might become painful to update that whenever there are changes, such as new error messages from new VK extensions.

renchenglei commented:
I have synced with Cactus team, when the build finished, the whole out folder will be deleted. Currently, we can't get 'libmesa_vulkan_util.a' and 'libmesa_vulkan_common.a'.
Yes, it will be painful for our rebase. Before we have a real fix, I will take care of this wa, we keep this wa to unblock the cactus building error, :)
We won't get this wa landed on our github, so opensource-gfx branch works as before. I will also keep investigating this issue with cactus and compile team to double confirm if this issue is from our side or Android Q toolchain.
